### ISSUE TITLE
Transition `_cents` fields to `display_.._dollars` fields (for non USD auctions)

### ISSUE THREAD

**Issue by mzikherman** [artsy/eigen#759]
In order to enable non USD auctions, recent PR's in gravity (https://github.com/artsy/gravity/pull/9179 as an example), has added JSON fields to help currency formatting.

So, any field relating to bidding that was of the form `_cents` (such as `estimate_cents`), now will also include a `display_estimate_dollars` field. This field (a string), will include any currency information, as well as proper formatting/rounding.

This makes it easier to actually put on a non USD auction, as clients can become 'dumber' and just render these fields directly, rather than doing a `divide by 100 and add a $ sign`, type thing.

This is fully backwards-compatible (it's just extra JSON) so no rush, but it would be nice (and I think net code deletion for you guys) to switch to this.

Here's an example of how Force was updated to use it: https://github.com/artsy/force/pull/3334

Will open this issue in Eidolon as well.

Thanks guys! Let me know if you have questions.

**Comment by alloy** [comment]
I like the idea of allowing clients to dumb down, but I’m worried about ignore the user’s locale preferences. For instance, my phone is configured to use Dutch number formatting, in which `$5,500` really means **five and a half dollars**.

While you could argue that most people interested in buying art –_certainly not most non-US/UK people_– will probably understand the formatting when the price is given in dollars, it’s not the best and welcoming UX.

I think that the best thing would be to do let the clients perform the formatting and also have them take the locale preferences into account. E.g. Force and Martsy would have to check what the user’s preferred locale is by checking the `Accept-Language` HTTP header and, in our case, by using an `NSNumberFormatter` everywhere.

Has there been any prior discussion about this?

----

Embarrassingly enough, Eigen is currently not respecting the user’s locale preferences in a consistent manner. For instance, when an artwork has been sold, a string version of the price is used (so apperantly something like this already exists), whereas with an ongoing auction, a NSNumberFormatter is used which automatically uses the preferred locale formatting.

### Incorrect version that uses a string from the API

![ios simulator screen shot 01 sep 2015 12 32 19](https://cloud.githubusercontent.com/assets/2320/9602841/303a4634-50ac-11e5-920c-41b6bd4b34ff.png)

### Correct version that uses client-side formatting

![ios simulator screen shot 01 sep 2015 13 14 56](https://cloud.githubusercontent.com/assets/2320/9602845/3f91d8cc-50ac-11e5-8e26-ac6e8d932ef6.png)

**Comment by alloy** [comment]
Holdup! It’s even _worse_ with the native view, as it is showing the price as being in euros, while the actual price is `$28,000`, meaning the artwork appears to be `$3539,25` more expensive than it actually is.

This is really bad :sob:

**Comment by alloy** [comment]
Filed #760 to deal with this and I’ll get on it immediately.

**Comment by mzikherman** [comment]
We haven't discussed non USD stuff except for wanting to be set up for it. We came quite close to putting on a GBP auction a little while ago!

For non USD priced artworks we handle things the exact same way - the server formats the price/dollars/cents/currency into consistent strings that can just be rendered by the client. It seemed like a reasonable approach for this since optically, a stumbling block to putting on a non USD auction was how various clients hard-coded USD assumptions around auctions.

To be honest, I have virtually zero knowledge about iOS devices, and this is the first I'm hearing about 'locale preferences'. Is that something in our app?

It still seems like we'd want to control upstream exactly how prices and currencies are displayed, IMO.

**Comment by mzikherman** [comment]
Is there any native bidding in Eigen? I wasn't sure, I thought just Eidolon needed this update but dropped the issue in here too.

**Comment by orta** [comment]
No native bidding, we pass straight through to martsy if auth'd

**Comment by alloy** [comment]
> To be honest, I have virtually zero knowledge about iOS devices, and this is the first I'm hearing about 'locale preferences'. Is that something in our app?

No it’s basically just localization in general (any platform will let you configure your locale preferences), this same issue applies to the web apps. Force and Martsy both show me prices using the wrong separators as well, which is why this still affects Eigen, even if it has no native bidding.

My point is that if you want to do the formatting right, you’ll have to take the user’s preferences into account, regardless of where you render the content (I’m fine with it being rendered on the server, if that’s deemed important).

While I use the English language on all my devices, I do not want to see a comma used as a thousands separator, regardless of the currency of a price I’m looking at. That’s just not how numbers are formatted here :) For example, see http://docs.oracle.com/cd/E19455-01/806-0169/overview-9/index.html

And since this is all related to preparing to do things right, I think it’s important to take these things into account. If it’s decided that prices are always going to be formatted as they would be in the country they’re from, then we’ll have to know that so that we can have it work consistently in Eigen.

**Comment by ashfurrow** [comment]
Hmm, yeah. I mean, usually it's an auction or a gallery or whoever who sets the price, and they want to set the unit, too. I think that makes sense, and that's what Matt has built. However, on iOS, we're displaying the prices using a locale-specific currency symbols and separators ([here](https://github.com/artsy/Artsy-UILabels/blob/6efceb550183915cde0c3acba042b52ea5322ade/Pod/Classes/NSNumberFormatter%2BARCurrency.m#L34-L52) is the code, which is causing the problem). 

Basically, I think the server should provide a symbol for the currency, a separator for each three digits ("1,000,000" vs "1 000 000"), and a decimal symbol for cents ("$1.50" vs "$1,50"). Does that make sense?

**Comment by mzikherman** [comment]
I really do feel like it's 'safer' or 'better' to not use localization preferences for this, as I have a suspicion the auction house/gallery will want to know exactly how things will look.

I can add to the JSON whatever fields are needed so that iOS devices can be more intelligent about formatting, if we feel that it is necessary.

For bidding, would the work be on Force/Martsy to support this, simply b/c Eigen is using web views for auctions? I'm not super experienced with supporting an iOS app like this- but it seems kind of burdensome and overcomplicated?? I dunno, that's just my initial opinion.

I'm more than happy to add fields to JSON or modify behavior in Force/Martsy - just let me know exactly what you need!

**Comment by ashfurrow** [comment]
> I really do feel like it's 'safer' or 'better' to not use localization preferences for this

Agreed! It should be up to the org selling the artwork to decide currency. Devang has talked about automatic currency conversion, but it's down the road a bit.

**Comment by mzikherman** [comment]
Discussed with @alloy offline, I'm much more knowledgable about this!

Interesting caveat - not only is the display different based on your preferences, but also your location! `$` vs `AU$` vs `US$`.

I still think this is a worthy update to have made in Force/Martsy (and Eidolon), but I understand the implications and what we would need to do to fully support this in a better way. I will raise the question at our next auctions meeting.

**Comment by ashfurrow** [comment]
@alloy I don't know if that's a safe assumption – I can see partners wanting complete control over how prices are displayed, without conversion to locale-specific currencies. An art gallery in NYC might not care about people in the UK seeing USD, as long as the price is accurate.

**Comment by alloy** [comment]
> An art gallery in NYC might not care about people in the UK seeing USD, as long as the price is accurate.

@ashfurrow That’s what I meant as well, as I’m not based in the UK ;) But my main point was, let’s not conflate these two things:

* The equal value of a price, regardless of currency (and thus automatic conversion to other currencies etc).
* How to format numbers in general, regardless of it being a price or any other number.

The latter is the only thing this thread was about, besides the bug I found in Eigen that looked as if it did automatic currency conversion.

**Comment by ashfurrow** [comment]
Agreed :100:

**Comment by alloy** [comment]
A summary of the possible strategies based on my talk with @mzikherman:

1. Go fully localized and let the user’s (locale) preferences dictate the number formatting. This can be done server or client side.
  - Pros: more readable to users that are accustomed to the number formatting rules in the country they grew up in, shows our dedication to international partners.
  - Cons: possibly hard to detect what the user’s preference is (e.g. do all browsers send the `Accept-Language` header?), harder to implement.
2. Have partners fully dictate how prices are formatted.
  - Pros: Partners feel in control, simple to implement.
  - Cons: As an international user looking at auctions in various countries (or even different partners in the same country), I could be shown numbers in various formats or even wonder what types of dollars a price is in.
3. Always use 1 format. E.g. US/UK number formatting.
  - Pros: consistent on all platforms, simple to implement.
  - Cons: less readable to users having grown up in a country where different formatting is used.

The web apps are somewhere between 2 and 3, whereas Eigen is a mixture of all of the above due to use of native and web views.

**Comment by devangt** [comment]
Thanks @ashfurrow.

Love the discussion so far! Sorry for jumping in a little late here. I have been trying to put some thought into this over the past week and landed at some similar feature areas. 

a) Partner settings: We need the creator of the auction to set which currency the auction is running in. This transcends to artwork price currencies and to bid strategies. This means a setting in Ohm and also in any future Auction CMS we create.

b) User Settings: A user may wish to set his/her global preferences to view everything in the currecy of their cosing (a few of our competitors allow this). However the currency of an auction will superseed this if we chose to; and if so we give the user a way to check the exchange rate estimate of a given piece when they are viewing it (aka on demand as @ashfurrow mentioned). This is something i have very initial sketches on and will share out a few screens this week and work with Katarina when shes back next week.

c) Final Billing: The billing always occurs in the currency that the Auction is running regardless of any setting and we need to make the user aware of that both in user ed and in our terms of use.

**Comment by ashfurrow** [comment]
And /cc @broskoski

**Comment by ashfurrow** [comment]
I think this ought to be moved to the Live milestone. @katarinabatina ?

**Comment by orta** [comment]
#1544 deals with this, not sure if it's _entirely_ covered, but definitely around live sales

**Comment by ashfurrow** [comment]
Issue moved to [artsy/auctions #196](https://github.com/artsy/auctions/issues/196) via [**ZenHub**](https://www.zenhub.com/)
